docs/proposals/20220803-rollouthistory.md

```diff
@@ -0,0 +1,332 @@
+	Name string `json:"name,omitempty"`
+	// Route service traffic to pods with label keys and values matching this
+	// selector.
+	// More info: https://kubernetes.io/docs/concepts/services-networking/service/
+	// +optional
+	// +mapType=atomic
+	Selector map[string]string `json:"selector,omitempty" protobuf:"bytes,2,rep,name=selector"`
+	// The list of ports that are exposed by this service.
+	// More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
+	// +patchMergeKey=port
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=port
+	// +listMapKey=protocol
+	Ports []v1.ServicePort `json:"ports,omitempty" patchStrategy:"merge" patchMergeKey:"port" protobuf:"bytes,1,rep,name=ports"`
+	// clusterIP is the IP address of the service and is usually assigned
+	// randomly.
+	// +optional
+	ClusterIP string `json:"clusterIP,omitempty" protobuf:"bytes,3,opt,name=clusterIP"`
+	// ClusterIPs is a list of IP addresses assigned to this service, and are
+	// usually assigned randomly.
+	// +optional
+	ClusterIPs []string `json:"clusterIPs,omitempty" protobuf:"bytes,18,opt,name=clusterIPs"`
+	// type determines how the Service is exposed.
+	// +optional
+	Type string `json:"type,omitempty" protobuf:"bytes,4,opt,name=type,casttype=ServiceType"`
+}
+
+// IngressInfo indicates information of the ingress
+type IngressInfo struct {
+	metav1.TypeMeta `json:",inline"`
+	// Name indicates the Ingress name
+	Name string `json:"name,omitempty"`
+	// DefaultBackend is the backend that should handle requests that don't
+	// match any rule. If Rules are not specified, DefaultBackend must be specified.
+	// If DefaultBackend is not set, the handling of requests that do not match any
+	// of the rules will be up to the Ingress controller.
+	// +optional
+	DefaultBackend *networkingv1.IngressBackend `json:"defaultBackend,omitempty" protobuf:"bytes,1,opt,name=defaultBackend"`
+	// A list of host rules used to configure the Ingress. If unspecified, or
+	// no rule matches, all traffic is sent to the default backend.
+	// +listType=atomic
+	// +optional
+	Rules []networkingv1.IngressRule `json:"rules,omitempty" protobuf:"bytes,3,rep,name=rules"`
+}
+
+// RolloutHistoryStatus defines the observed state of RolloutHistory
+type RolloutHistoryStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Status indicates state of RolloutHistory, such as "progressing", "completed", "terminated"
+	Status string `json:"status,omitempty"`
+	// CurrentStep indicates the current step
+	CurrentStep *int32 `json:"currentstep,omitempty"`
+	// RHPhase indicates state of this rollout revision, such as "init", "pending", "update", "terminated", "completed".
+	RHPhase string `json:"rhphase,omitepty"`
+	// RolloutStatus indicates the rollouts status
+	RolloutStatus RolloutStatusInRH `json:"rolloutStatus,omitempty"`
```

Comment:
We don't ` need RolloutStatus `  to record the process status of rollout, because the rollout will finish any way.